File: Sources/Olaf/UI/Util/CurlBuilder.swift
import Foundation

/// Produces a cURL representation from a `.network` entry. Header/body values are **raw** (no
/// masking); secrets, including `Authorization`/`Cookie`, pass through verbatim into the output —
/// review before sharing.
enum CurlBuilder {
    static func curl(from info: NetworkLogInfo) -> String {
        var lines = ["curl -X \(info.method ?? "GET") \(quote(info.url ?? ""))"]
        for header in info.requestHeaders {
            lines.append("-H \(quote("\(header.key): \(header.value)"))")
        }
        if let body = info.requestBody, !body.isEmpty {
            lines.append("-d \(quote(body))")
        }
        return lines.joined(separator: " \\\n  ")
    }

    /// Single-quoted shell argument; safely escapes any embedded single quotes.
    private static func quote(_ value: String) -> String {
        "'" + value.replacingOccurrences(of: "'", with: "'\\''") + "'"
    }
}

File: Android/olaf/src/main/java/com/olaf/ui/util/CurlBuilder.kt
package com.olaf.ui.util

import com.olaf.ui.model.NetworkLogInfo

/**
 * Renders a captured request as a cURL command.
 *
 * Header and body values are **raw** — secrets such as `Authorization` and `Cookie` pass through
 * verbatim, which is exactly what makes the output useful and exactly why it must be reviewed
 * before being shared.
 */
internal object CurlBuilder {

    fun curl(info: NetworkLogInfo): String {
        val parts = mutableListOf("curl -X ${info.method ?: "GET"} ${quote(info.url.orEmpty())}")
        info.requestHeaders.forEach { (key, value) ->
            parts.add("-H ${quote("$key: $value")}")
        }
        info.requestBody?.takeIf { it.isNotEmpty() }?.let { parts.add("-d ${quote(it)}") }
        return parts.joinToString(" \\\n  ")
    }

    /** Single-quoted shell argument, with embedded single quotes escaped safely. */
    private fun quote(value: String): String = "'" + value.replace("'", "'\\''") + "'"
}
